builtin/stash--helper.c

```diff
@@ -0,0 +1,70 @@
+#include "../cache.h"
+#include "../cache-tree.h"
+#include "../diff.h"
+#include "../diffcore.h"
+#include "../revision.h"
+
+static const char builtin_stash__helper_usage[] = {
+	"Usage: git stash--helper --non-patch <i_tree>"
+};
+
+static void diff_callback(struct diff_queue_struct *q,
+					 struct diff_options *options, void *data)
+{
+	int i;
+	for (i = 0; i < q->nr; i++)
+    {
+		struct diff_filepair *p;
+		struct stat st;
+
+		p = q->queue[i];
+		if (p->one && lstat(p->one->path, &st)) {
+			warning("Could not stat %s", p->one->path);
+			continue;
+		}
+		add_to_index(&the_index, p->one->path, &st, 0);
+	}
+}
+
+int stash_non_patch(const char *i_tree,	const char *prefix)
+{
+	int result;
+	struct rev_info rev;
+	const char *wt_prefix = NULL;
+	unsigned char sha1[20];
+	const char *me = "git stash--helper";
+
+	read_index(&the_index);
+    refresh_cache(REFRESH_QUIET);
+	init_revisions(&rev, NULL);
+	add_head_to_pending(&rev);
+	rev.diffopt.output_format |= DIFF_FORMAT_CALLBACK;
+	rev.diffopt.format_callback = diff_callback;
+	rev.diffopt.format_callback_data=&result;
+	diff_setup_done(&rev.diffopt);
+    run_diff_index(&rev,0);
+    result = write_cache_as_tree(sha1, 0, wt_prefix);
```

Comment:
Yeah, I only did a quick check for line endings. Every other formating issue is considered a known issue.